code/CloudFlare_Purge.php

```diff
@@ -0,0 +1,433 @@
+     *
+     * @param     $pattern
+     * @param int $flags
+     *
+     * @return array
+     */
+    private function rglob($pattern, $flags = 0)
+    {
+        $files = glob($pattern, $flags);
+        foreach (glob(dirname($pattern) . '/*', GLOB_ONLYDIR | GLOB_NOSORT) as $dir) {
+            $files = array_merge($files, $this->rglob($dir . '/' . basename($pattern), $flags));
+        }
+
+        return $files;
+    }
+
+    /**
+     * Converts /public_html/path/to/file.ext to example.com/path/to/file.ext
+     *
+     * @param string|array $files
+     *
+     * @return mixed
+     */
+    public function convertToAbsolute($files)
+    {
+        // It's not the best feeling to have to add http:// here, despite it's SSL variant being picked up
+        // by getUrlVariants(). However without it cloudflare will response with an error similiar to:
+        // "You may only purge files for this zone only"
+        $baseUrl = "http://" . CloudFlare::singleton()->getServerName() . "/";
+        $rootDir = $_SERVER['DOCUMENT_ROOT'] . Director::baseURL();
+
+        if (is_array($files)) {
+            foreach ($files as $index => $file) {
+                $basename = basename($file);
+                $basenameEncoded = urlencode($basename);
+                $file = str_replace($basename, $basenameEncoded, $file);
+
+                $files[$index] = str_replace($rootDir, $baseUrl, $file);
+            }
+
+            return $files;
+        }
+
+        if (is_string($files)) {
+            $basename = basename($files);
+            $basenameEncoded = urlencode($basename);
+            $files = str_replace($basename, $basenameEncoded, $files);
+
+            return str_replace($rootDir, $baseUrl, $files);
+        }
+
+        return false;
+    }
+
+    /**
+     * @return int
+     */
+    public function count()
+    {
+        return (is_array($this->files)) ? count($this->files) : 0;
+    }
+
+    /**
+     *
+     */
+    public function purge()
+    {
+        $files = $this->getFiles();
+
+        $this->extend("updateFilesBeforePurge", $files);
+
+        if ($this->purgeEverything) {
+            $data = array(
+                "purge_everything" => true
+            );
+        } else {
+            $data = array(
+                "files" => $files
+            );
+        }
+
+        $this->setResponse($this->handleRequest($data));
+
+        $success = $this->isSuccessful();
+
+        if ($success && $this->getSuccessMessage()) {
+            CloudFlare_Notifications::handleMessage(
+                $this->getSuccessMessage(),
+                array(
+                    'file_count' => $this->count()
+                )
+            );
+        }
+
+        if (!$success && $this->getFailureMessage()) {
+            CloudFlare_Notifications::handleMessage(
+                $this->getFailureMessage(),
+                array(
+                    'file_count' => $this->count()
+                )
+            );
+        }
+
```

Comment:
This seems like it should be one method call with one message, they're basically identical